File: sdks/python/nroutersdk/prompts.py
"""Managed prompt helpers for the nRouter request fields."""

from __future__ import annotations

import re
from collections.abc import Mapping
from dataclasses import dataclass

from nroutersdk._errors import nRouterRequestError
from nroutersdk._options import (
    PROMPT_TEMPLATE_ID_FIELD,
    PROMPT_VARIABLES_FIELD,
    build_extra_body,
)

PROMPT_WIRE_FIELDS = (PROMPT_TEMPLATE_ID_FIELD, PROMPT_VARIABLES_FIELD)
SYSTEM_VARIABLE_NAMES = ("org_name", "model", "timestamp", "user_id")


@dataclass(frozen=True)
class PromptSelection:
    """One request's prompt selection."""

    template_id: str | None = None
    variables: dict[str, str] | None = None


def prompt_template(
    template_id: str, variables: Mapping[str, str] | None = None
) -> PromptSelection:
    """Select a specific prompt template for one request."""
    if not isinstance(template_id, str) or not template_id.strip():
        raise nRouterRequestError(
            "prompt_template requires a template id. Pass the template id, or "
            "call prompt_variables() to use the assignment for this key, team, "
            "or organization."
        )
    return PromptSelection(
        template_id=template_id.strip(),
        variables=dict(variables) if variables is not None else None,
    )


def prompt_variables(variables: Mapping[str, str]) -> PromptSelection:
    """Render the prompt already assigned to this key, team, or organization."""
    return PromptSelection(variables=dict(variables))


def with_variables(selection: PromptSelection, more: Mapping[str, str]) -> PromptSelection:
    """Return a new selection with extra variables, later values winning."""
    return PromptSelection(
        template_id=selection.template_id,
        variables={**(selection.variables or {}), **dict(more)},
    )


def prompt_extra_body(selection: PromptSelection) -> dict[str, object]:
    """Map a prompt selection to nRouter's request body fields."""
    return build_extra_body(
        prompt_template_id=selection.template_id,
        prompt_variables=selection.variables,
    )


def apply_prompt(options: dict[str, object], selection: PromptSelection) -> dict[str, object]:
    """Return a new options dict with the prompt selection applied."""
    next_options = dict(options)
    if selection.template_id:
        next_options["prompt_template_id"] = selection.template_id
    if selection.variables:
        existing = next_options.get("prompt_variables")
        merged = dict(existing) if isinstance(existing, (dict, Mapping)) else {}
        merged.update(selection.variables)
        next_options["prompt_variables"] = merged
    return next_options


def system_variable_conflicts(variables: Mapping[str, str] | None) -> list[str]:
    """Names the caller supplied that the gateway will overwrite."""
    if not variables:
        return []
    return [name for name in SYSTEM_VARIABLE_NAMES if name in variables]


_VARIABLE_PATTERN = re.compile(r"\{\{\s*([a-zA-Z0-9_-]+)\s*\}\}")


def render_prompt(
    template: str,
    variables: Mapping[str, object] | None = None,
    *,
    strict: bool = False,
    system_variables: Mapping[str, str] | None = None,
) -> str:
    """Safely renders a prompt template by interpolating `{{variable}}` tokens.

    Security & resiliency properties:
    - Single-pass replacement: prevents recursive expansion loops.
    - Callable replacer: prevents regex backreference escapes (\\1, \\g<...>).
    - Strict mode: raises nRouterRequestError if any template variable is missing.
    - System variables: take precedence over caller variables matching gateway rules.
    """
    if not isinstance(template, str):
        return ""

    missing_keys: list[str] = []

    def _repl(match: re.Match[str]) -> str:
        key = match.group(1)
        if system_variables and key in system_variables:
            sys_val = system_variables[key]
            return "" if sys_val is None else str(sys_val)
        if variables and key in variables:
            user_val = variables[key]
            return "" if user_val is None else str(user_val)
        if strict:
            missing_keys.append(key)
        return match.group(0)

    result = _VARIABLE_PATTERN.sub(_repl, template)
    if strict and missing_keys:
        raise nRouterRequestError(
            f"Missing required prompt template variables: {', '.join(missing_keys)}"
        )
    return result

File: sdks/go/prompts.go
package nrouter

import (
	"fmt"
	"regexp"
	"strings"
)

const (
	PromptTemplateIDField = "nrouter_prompt_template_id"
	PromptVariablesField  = "nrouter_prompt_variables"
)

var (
	PromptWireFields    = []string{PromptTemplateIDField, PromptVariablesField}
	SystemVariableNames = []string{"org_name", "model", "timestamp", "user_id"}
)

// PromptSelection holds prompt template configuration.
type PromptSelection struct {
	TemplateID string
	Variables  map[string]any
}

// PromptTemplate creates a PromptSelection specifying a template ID and optional variables.
func PromptTemplate(id string, variables ...map[string]any) (PromptSelection, error) {
	trimmed := strings.TrimSpace(id)
	if trimmed == "" {
		return PromptSelection{}, fmt.Errorf("prompt_template requires a non-empty template id")
	}
	vars := make(map[string]any)
	if len(variables) > 0 && variables[0] != nil {
		for k, v := range variables[0] {
			vars[k] = v
		}
	}
	return PromptSelection{
		TemplateID: trimmed,
		Variables:  vars,
	}, nil
}

// PromptVariables creates a PromptSelection specifying only variables for the assigned template.
func PromptVariables(variables map[string]any) PromptSelection {
	vars := make(map[string]any)
	if variables != nil {
		for k, v := range variables {
			vars[k] = v
		}
	}
	return PromptSelection{
		Variables: vars,
	}
}

// WithVariables returns a new PromptSelection with merged variables (incoming overrides).
func (p PromptSelection) WithVariables(variables map[string]any) PromptSelection {
	merged := make(map[string]any)
	for k, v := range p.Variables {
		merged[k] = v
	}
	if variables != nil {
		for k, v := range variables {
			merged[k] = v
		}
	}
	return PromptSelection{
		TemplateID: p.TemplateID,
		Variables:  merged,
	}
}

// ApplyTo adds prompt template fields to an extra body map.
func (p PromptSelection) ApplyTo(extra map[string]any) {
	if extra == nil {
		return
	}
	if p.TemplateID != "" {
		extra[PromptTemplateIDField] = p.TemplateID
	}
	if len(p.Variables) > 0 {
		extra[PromptVariablesField] = p.Variables
	}
}

// SystemVariableConflicts checks if any caller variables collide with gateway system variables.
// Returns conflicting variable names in the gateway's deterministic insertion order.
func SystemVariableConflicts(variables map[string]any) []string {
	if variables == nil {
		return nil
	}
	var conflicts []string
	for _, sysVar := range SystemVariableNames {
		if _, exists := variables[sysVar]; exists {
			conflicts = append(conflicts, sysVar)
		}
	}
	return conflicts
}

// RenderPromptOptions controls client-side prompt template interpolation.
type RenderPromptOptions struct {
	Strict          bool
	SystemVariables map[string]string
}

var promptVariableRegex = regexp.MustCompile(`\{\{\s*([a-zA-Z0-9_-]+)\s*\}\}`)

// RenderPrompt safely interpolates `{{variable}}` or `{{ variable }}` tokens in a template.
//
// Security & resiliency features:
// - Single-pass replacement prevents recursive variable expansion loops.
// - Safe interpolation: no metacharacter injection from variable values.
// - Strict mode: returns error if any template variable is missing.
// - System variables: take precedence over caller variables matching gateway rules.
func RenderPrompt(template string, variables map[string]any, opts ...RenderPromptOptions) (string, error) {
	if template == "" {
		return "", nil
	}
	var options RenderPromptOptions
	if len(opts) > 0 {
		options = opts[0]
	}

	var missingKeys []string
	result := promptVariableRegex.ReplaceAllStringFunc(template, func(match string) string {
		submatches := promptVariableRegex.FindStringSubmatch(match)
		if len(submatches) < 2 {
			return match
		}
		key := submatches[1]

		// 1. System variables take precedence
		if options.SystemVariables != nil {
			if sysVal, ok := options.SystemVariables[key]; ok {
				return sysVal
			}
		}

		// 2. Caller variables
		if variables != nil {
			if val, ok := variables[key]; ok {
				if val == nil {
					return ""
				}
				return fmt.Sprintf("%v", val)
			}
		}

		if options.Strict {
			missingKeys = append(missingKeys, key)
		}
		return match
	})

	if options.Strict && len(missingKeys) > 0 {
		return "", fmt.Errorf("missing required prompt template variables: %s", strings.Join(missingKeys, ", "))
	}

	return result, nil
}
